test/rollup_circuit.js

```diff
@@ -453,65 +608,1056 @@ describe("Rollup Basic circuit TXs", function () {
-    it("Should check underflow onchain", async () => {
-    });
-    it("Should check underflow offchain", async () => {
-    });
-    it("Should check offchain with loadAmount", async () => {
-    });
-    it("Should check onchain transfer", async () => {
-    });
-    it("Should check onchain deposit existing", async () => {
-    });
-    it("Should check combined deposit transfer", async () => {
-    });
-    it("Should check combined deposit exit", async () => {
-    });
-    it("Should check offchain with invalid fee", async () => {
-    });
-    it("Should check offchain with invalid nonce", async () => {
-    });
-    it("Should check offchain with invalid signature", async () => {
-    });
-    it("Should check offchain with not defined coin", async () => {
-    });
-    it("Should check offchain with double defined coin", async () => {
-    });
-    it("Should check batch with invalid order", async () => {
+    it("Should check underflow onchain", async () => { 
+        // Start a new state
+        const db = new SMTMemDB();
+        const rollupDB = await RollupDB(db);
+        const bb = await rollupDB.buildBatch(NTX, NLEVELS);
+  
+        const account1 = new RollupAccount(1);
+        const account2 = new RollupAccount(2);
+  
+        bb.addTx({
+            fromIdx: 1,
+            loadAmount: 1000,
+            coin: 0,
+            ax: account1.ax,
+            ay: account1.ay,
+            ethAddress: account1.ethAddress,
+            onChain: true
+        });
+  
+        bb.addTx({
+            fromIdx: 2,
+            loadAmount: 2000,
+            coin: 0,
+            ax: account2.ax,
+            ay: account2.ay,
+            ethAddress: account2.ethAddress,
+            onChain: true
+        });
+  
+        await bb.build();
+        const input = bb.getInput();
+  
+        const w = circuit.calculateWitness(input, {logTrigger:false, logOutput: false, logSet: false});
+        checkBatch(circuit, w, bb);
+  
+        await rollupDB.consolidate(bb);
+  
+        const bb2 = await rollupDB.buildBatch(NTX, NLEVELS);
+  
+        const tx = {
+            fromIdx: 1,
+            toIdx: 2,
+            coin: 0,
+            amount: 5000,
+            nonce: 0,
+            ax: account1.ax,
+            ay: account1.ay,
+            ethAddress: account1.ethAddress,
+            onChain: true
+        };
+        bb2.addTx(tx);
+
+        await bb2.build();
+        const input2 = bb2.getInput();
+            
+        const w2 = circuit.calculateWitness(input2, {logTrigger:false, logOutput: false, logSet: false});
+        checkBatch(circuit, w2, bb2);
+        await rollupDB.consolidate(bb2);
+        const s2_1 = await rollupDB.getStateByIdx(1);
+        const s2_2 = await rollupDB.getStateByIdx(2);
+        assert.equal(s2_1.amount.toString(), 1000);
+        assert.equal(s2_2.amount.toString(), 2000);
+        
+    });
+    it("Should check onchain deposit of 128 bits", async () => { // should SC prevent this
```

Comment:
Smart contract does not allow deposits above 2**128 bits, then a deposit below this amount is a regular deposit which has been covered in previous test